core/src/test/java/io/parsingdata/metal/data/ParseValueCacheTest.java

```diff
@@ -0,0 +1,173 @@
+        final ParseValue pv1 = parseValue("name");
+        final ParseValueCache newCache = noCache.add(pv1);
+        assertFalse(newCache.find("name", NO_LIMIT).isPresent());
+        assertEquals(NO_CACHE, newCache);
+    }
+
+    @Test
+    public void addTest() {
+        final ParseValueCache parseValueCache = new ParseValueCache();
+        final ParseValue pv = parseValue("name");
+        final ParseValueCache parseValueCache2 = parseValueCache.add(pv);
+
+        final Optional<ImmutableList<Value>> nameValues = parseValueCache.find("name", NO_LIMIT);
+        assertTrue(nameValues.isPresent());
+        nameValues.ifPresent(result -> assertEquals(0, result.size));
+
+        final Optional<ImmutableList<Value>> nameValues2 = parseValueCache2.find("name", NO_LIMIT);
+        assertTrue(nameValues2.isPresent());
+        nameValues2.ifPresent(result -> {
+            assertEquals(1, result.size);
+            assertEquals(pv, result.head);
+        });
+    }
+
+    @Test
+    public void toStringTest() {
+        assertEquals("cache:size=4", parseValueCache.toString());
+        assertEquals("no-cache", NO_CACHE.toString());
+    }
+
+    public static Stream<Arguments> cacheUsageTest() {
+        return Stream.of(
+            arguments("nameRef", ref("second.name"), true),
+            arguments("nameRef with limit", ref(con(1), "second.name"), true),
+
+            arguments("multi nameRef", ref("second.name", "first.name"), false),
+            arguments("multi nameRef with limit", ref(con(1), "second.name", "first.name"), false),
+            arguments("definitionRef", ref(pv2Definition), false),
+            arguments("definitionRef with limit", ref(con(1), pv2Definition), false),
+            arguments("multi definitionRef", ref(pv2Definition, pv3Definition), false),
+            arguments("multi definitionRef with limit", ref(con(1), pv2Definition, pv3Definition), false),
+
+            arguments("scoped nameRef", scope(ref("second.name"), con(1)), false),
+            arguments("scoped nameRef with limit", scope(ref(con(1), "second.name"), con(1)), false),
+            arguments("scoped multi nameRef", scope(ref("second.name", "first.name"), con(1)), false),
+            arguments("scoped multi nameRef with limit", scope(ref(con(1), "second.name", "first.name"), con(1)), false),
+            arguments("scoped definitionRef", scope(ref(pv2Definition), con(1)), false),
+            arguments("scoped definitionRef with limit", scope(ref(con(1), pv2Definition), con(1)), false),
+            arguments("scoped multi definitionRef", scope(ref(pv2Definition, pv3Definition), con(1)), false),
+            arguments("scoped multi definitionRef with limit", scope(ref(con(1), pv2Definition, pv3Definition), con(1)), false)
+        );
+    }
+
+    @ParameterizedTest(name="{2} - {0}")
+    @MethodSource
+    public void cacheUsageTest(final String testName, final ValueExpression expression, final boolean shouldUseCache) {
```

Comment:
This tests if the cache is used. The other tests within this class make sure that the cache works as expected. Other tests outside this class, tests if the results are expected. This tests shows in which cases the cache is actually used.